tests/test_fts3_transfer_tools.py

```diff
@@ -0,0 +1,397 @@
+    config_get,
+    config_remove_option,
+)
+from rucio.db.sqla.session import get_session
+from rucio.daemons.conveyor.submitter import submitter
+from rucio.core.plugins import PolicyPackageAlgorithms
+from rucio.transfertool.fts3_plugins import FTS3SchedulingHints, FTS3CollocationHints
+
+from rucio.core import distance as distance_core
+from rucio.core import replica as replica_core
+from rucio.core import rule as rule_core
+
+
+mock_session = get_session()
+
+MAX_POLL_WAIT_SECONDS = 60
+TEST_FTS_HOST = "https://fts:8446"
+
+config_add_section("transfers")
+config_add_section("tape_prority")
+
+
+class FTS3CollocationHintsMock(FTS3CollocationHints):
+    def __init__(self):
+        super().__init__()
+
+    def set_collocation(self, *args, **kwargs):
+        return {"0": "", "1": "", "2": "", "3": ""}
+
+
+def _make_config_section():
+    try:
+        config_add_section("transfers")
+        config_add_section("tape_prority")
+    except:
+        pass
+
+
+def _make_transfer_path(did, rse_factory, root_account):
+    _, src_rse_id = rse_factory.make_mock_rse()
+    dst_rse, dst_rse_id = rse_factory.make_mock_rse()
+    all_rses = [src_rse_id, dst_rse_id]
+    distance_core.add_distance(src_rse_id, dst_rse_id, distance=1)
+
+    topology = Topology(rse_ids=all_rses)
+    file = {"scope": did["scope"], "name": did["name"], "bytes": 1}
+    replica_core.add_replicas(rse_id=src_rse_id, files=[file], account=root_account)
+    rule_core.add_rule(
+        dids=[did],
+        account=root_account,
+        copies=1,
+        rse_expression=dst_rse,
+        grouping="ALL",
+        weight=None,
+        lifetime=None,
+        locked=False,
+        subscription_id=None,
+    )
+
+    requests_by_id = list_and_mark_transfer_requests_and_source_replicas(
+        rse_collection=topology, rses=[src_rse_id, dst_rse_id]
+    )
+    requests, *_ = build_transfer_paths(
+        topology=topology,
+        protocol_factory=ProtocolFactory(),
+        requests_with_sources=requests_by_id.values(),
+    )
+    _, [transfer_path] = next(iter(requests.items()))
+
+    return transfer_path
+
+
+def test_scheduling_hints(did_factory, rse_factory, root_account):
+
+    """Transfer with an installed plugin for scheduling based on transfer activity"""
+
+    # Produce a new did
+    mock_did = did_factory.random_file_did()
+
+    # Make the transfer path
+    transfer_path = _make_transfer_path(mock_did, rse_factory, root_account)
+
+    activity = transfer_path[0].rws.activity
+    _make_config_section()
+    config_set(section="tape_prority", option=activity, value="100")
+    assert config_get("tape_prority", activity) == "100"
+
+    # Inject plugin into existing config
+    config_set(
+        section="transfers",
+        option="plugins",
+        value="[FTS3SchedulingHints]",
+    )
+    FTS3SchedulingHints()
+
+    # Mock Transfer Tool
+    fts3_tool = FTS3Transfertool(TEST_FTS_HOST)
+    assert "FTS3SchedulingHints" in fts3_tool.plugins
```

Comment:
I, personally, am a  very strong opponent of tests which verify the status of internal data structures. If we rename self.plugins to self.metadata_plugins later, we have to also update tests. Any simple refactoring will require updating the tests. This way, if you introduce bugs, you end also updating the test to fit the new bugs. Making the test useless. I advise to update some tests in `test_conveyor.py` by setting the correct configuration values using file_config_mock and core_config_mock fixtures. Following by checking that FTS (not rucio!) does what it's intended to do.

Reply:
Honestly it's just how I was taught to do tests and sort of did it uncritically, but your logic makes sense. They also end up being redundant, so, removed. I was struggling with figuring out how to verify fts is correctly using the metadata, but I will check out `test_conveyor`, thanks!

Reply:
So I did start adding conveyor tests, but I consistently get an error `root:fts3.py:1015 Could not submit transfer to https://fts:8446 - HTTPSConnectionPool(host='fts', port=8446): Max retries exceeded with url: /jobs (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f0e9c2d36d0>: Failed to establish a new connection: [Errno -2] Name or service not known'))`
Is there is a known work-around or known way to properly connect to that fts server that I'm missing?

Reply:
How are you running the tests ? You need to run them from within the docker-compose env: https://rucio.github.io/documentation/operator/setting_up_demo/#using-the-environment-including-storage 
(also see the previous section on how to run the tests). If you want to use it from a GUI and have a paid intellij licence, you can use the quite neat integration (only tested on linux): https://rucio.github.io/documentation/developer/setting_up_intellij_dev_env; or (a bit more complex to set-up, but free) vscode alternative: https://rucio.github.io/documentation/developer/setting_up_vscode_dev_env

Reply:
Oh, yeah I've been using the vscode integration, no worries. Never mind though, looks like I was just mis-reading some of the error messages.